File: modules/base/logger/src/main/java/candle/logger/Logger.java
package candle.logger;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class Logger {
  // Single Thread Executor for all logging operations
  private static final ExecutorService executor = Executors.newSingleThreadExecutor();
  private static final LoggingStrategy debugStrategy = new CandleLoggingStrategy(
          "debug", "DEBUG", executor, ANSIColors.GREEN);
  private static final LoggingStrategy errorStrategy = new CandleLoggingStrategy(
          "error", "ERROR", executor, ANSIColors.RED);
  private static final LoggingStrategy fatalStrategy = new CandleLoggingStrategy(
          "error", "FATAL", executor, ANSIColors.BOLD_RED);
  private static final LoggingStrategy infoStrategy = new CandleLoggingStrategy(
          "info", "INFO", executor, ANSIColors.CYAN);
  private static final LoggingStrategy stackTraceStrategy = new StackTraceLoggingStrategy(executor);
  private static final LoggingStrategy warnStrategy = new CandleLoggingStrategy(
          "info", "WARN", executor, ANSIColors.YELLOW);
  private final String name; // Name of the Module, Plugin and so on.
  private final boolean debug; // Is the Logger in Debug Mode.

  public Logger() throws
                  IOException {
    this(false);
  }

  public Logger( boolean debug, String name ) throws
                                              IOException {
    this.name = name;
    this.debug = debug;

    Files.createDirectories(Path.of(System.getProperty("user.dir") + "/logs"));
  }

  public Logger( boolean debug ) throws
                                 IOException {
    this(debug, "CandleMC");
  }

  public void log( LoggingStrategy strategy, Object... content ) {
    String FORMAT_NORMAL = "[%timestamp] - %log_level - %name - %content";
    String FORMAT_DEBUG = "[%timestamp] - %log_level - %name - %package:%line - %content";
    strategy.log(
            this.name,
            debug ? FORMAT_DEBUG : FORMAT_NORMAL,
            System.currentTimeMillis(),
            content);
  }

  public void debug( Object... content ) {
    log(debugStrategy, content);
  }

  public void error( Object... content ) {
    log(errorStrategy, content);
  }

  public void fatal( Object... content ) {
    log(fatalStrategy, content);
  }

  public void info( Object... content ) {
    log(infoStrategy, content);
  }

  public void stacktrace( Throwable stackTrace ) {
    log(stackTraceStrategy, stackTrace);
  }

  public void warn( Object... content ) {
    log(warnStrategy, content);
  }
}

File: modules/base/logger/src/main/kotlin/candle.logger/Logger.kt
package candle.logger

import java.nio.file.Files
import java.nio.file.Path
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class Logger @JvmOverloads constructor(// Is the Logger in Debug Mode.
    private val debug: Boolean = false, // Name of the Module, Plugin and so on.
    private val name: String = "CandleMC"
) {
    init {
        Files.createDirectories(Path.of(System.getProperty("user.dir") + "/logs"))
    }

    fun log(strategy: LoggingStrategy, vararg content: Any) {
        val FORMAT_NORMAL = "[%timestamp] - %log_level - %name - %content"
        val FORMAT_DEBUG = "[%timestamp] - %log_level - %name - %package:%line - %content"
        strategy.log(
            this.name,
            if (debug) FORMAT_DEBUG else FORMAT_NORMAL,
            System.currentTimeMillis(),
            *content
        )
    }

    fun debug(vararg content: Any) {
        log(debugStrategy, *content)
    }

    fun error(vararg content: Any) {
        log(errorStrategy, *content)
    }

    fun fatal(vararg content: Any) {
        log(fatalStrategy, *content)
    }

    fun info(vararg content: Any) {
        log(infoStrategy, *content)
    }

    fun stacktrace(stackTrace: Throwable) {
        log(stackTraceStrategy, stackTrace)
    }

    fun warn(vararg content: Any) {
        log(warnStrategy, *content)
    }

    companion object {
        // Single Thread Executor for all logging operations
        private val executor: ExecutorService = Executors.newSingleThreadExecutor()
        private val debugStrategy: LoggingStrategy = CandleLoggingStrategy(
            "debug", "DEBUG", executor, ANSIColors.GREEN
        )
        private val errorStrategy: LoggingStrategy = CandleLoggingStrategy(
            "error", "ERROR", executor, ANSIColors.RED
        )
        private val fatalStrategy: LoggingStrategy = CandleLoggingStrategy(
            "error", "FATAL", executor, ANSIColors.BOLD_RED
        )
        private val infoStrategy: LoggingStrategy = CandleLoggingStrategy(
            "info", "INFO", executor, ANSIColors.CYAN
        )
        private val stackTraceStrategy: LoggingStrategy = StackTraceLoggingStrategy(executor)
        private val warnStrategy: LoggingStrategy = CandleLoggingStrategy(
            "info", "WARN", executor, ANSIColors.YELLOW
        )
    }
}
